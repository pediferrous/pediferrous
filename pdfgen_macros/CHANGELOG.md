# Changelog

## 0.1.0 (2025-08-13)


### Features

* implement font support ([#52](https://github.com/pediferrous/pediferrous/issues/52)) ([92cdce0](https://github.com/pediferrous/pediferrous/commit/92cdce0589be53378db1ff24296a603ceb2a3490)), closes [#37](https://github.com/pediferrous/pediferrous/issues/37)
