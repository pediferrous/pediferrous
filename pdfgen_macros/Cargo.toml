[package]
name = "pdfgen_macros"
version.workspace = true
edition.workspace = true
authors.workspace = true
description.workspace = true
repository.workspace = true
homepage.workspace = true
readme.workspace = true
license.workspace = true

[lib]
proc-macro = true

[dependencies]
syn = {version = "2.0", features = ["parsing"]}
quote = "1.0"
