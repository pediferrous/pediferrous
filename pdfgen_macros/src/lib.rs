use proc_macro::TokenStream;

mod name;

#[proc_macro]
pub fn const_names(token_stream: TokenStream) -> TokenStream {
    name::const_names(token_stream)
}
