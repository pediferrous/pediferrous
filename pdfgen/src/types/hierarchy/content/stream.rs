use std::io::{Error, Write};

use pdfgen_macros::const_names;

use crate::types::{
    constants,
    hierarchy::primitives::{name::Name, obj_id::ObjId},
};

/// A stream object, like a string object, is a sequence of bytes that may be of unlimited length.
/// Streams should be used to represent objects with potentially large amounts of data, such as
/// images and page descriptions.
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct Stream {
    /// Object ID of this `Stream`.
    pub(crate) id: ObjId,

    // NOTE: Stream dictionaries have more entries such as filter, decode parameters etc. For now,
    //       we only need the required dictionary entry 'Length', implicitly available in `Vec`
    //       implementation.
    // TODO: Implement full support for stream dictionary.
    /// Bytes contained in this `Stream` object.
    inner: Vec<u8>,
}

impl Stream {
    const START_STREAM: &[u8] = b"stream";
    const END_STREAM: &[u8] = b"endstream";
    const_names!(LENGTH);

    /// Creates a new empty `Stream`, containing no bytes and with length 0.
    pub fn new(id: ObjId) -> Self {
        Self {
            id,
            inner: Vec::default(),
        }
    }

    /// Creates a new `Stream` with given bytes as the stream's bytes.
    pub fn with_bytes(id: ObjId, bytes: impl Into<Vec<u8>>) -> Self {
        Self {
            id,
            inner: bytes.into(),
        }
    }

    /// Writes (aditional) bytes into this `Stream`, updating it's length.
    pub fn push_bytes(&mut self, bytes: &[u8]) {
        self.inner
            .write_all(bytes)
            .expect("Writing to Vec should never fail.");
    }

    /// Write the stream object into the given implementor of [`Write`] trait, with dictionary
    /// containing only the required `Length` field.
    #[inline(always)]
    pub fn write(&self, writer: &mut dyn Write) -> Result<usize, Error> {
        self.write_with_dict(writer, |_| Ok(0))
    }

    /// Write the stream object into the given implementor of [`Write`] trait, with function that
    /// writes dictionary fields additional to the `Length` field.
    pub fn write_with_dict<F>(&self, writer: &mut dyn Write, write_dict: F) -> Result<usize, Error>
    where
        F: FnOnce(&mut dyn Write) -> Result<usize, Error>,
    {
        let mut written = pdfgen_macros::write_chain! {
            // BEGIN_DICTIONARY:
            writer.write(b"<< "),
            // write the additional dictionary fields
            write_dict(writer),

            // write the length
            Self::LENGTH.write(writer),
            writer.write(self.inner.len().to_string().as_bytes()),
            writer.write(b" >>"),
            writer.write(constants::NL_MARKER),
            // END_DICTIONARY

            // stream
            writer.write(Self::START_STREAM),
            writer.write(constants::NL_MARKER),
        };

        writer.write_all(&self.inner)?;
        written += self.inner.len();

        written += pdfgen_macros::write_chain! {
            writer.write(constants::NL_MARKER),
            writer.write(Self::END_STREAM),
        };

        Ok(written)
    }

    /// Returns `true` if no bytes were written to this [`Stream`].
    pub fn is_empty(&self) -> bool {
        self.inner.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use crate::types::hierarchy::primitives::obj_id::IdManager;

    use super::Stream;

    #[test]
    fn basic_stream() {
        let bytes = String::from("This is the content of a stream.");

        let mut id_manager = IdManager::default();
        let stream = Stream::with_bytes(id_manager.create_id(), bytes);

        let mut writer = Vec::default();
        stream.write(&mut writer).unwrap();
        let output = String::from_utf8_lossy(&writer);

        insta::assert_snapshot!(output, @r"
        << /Length 32 >>
        stream
        This is the content of a stream.
        endstream
        ");
    }
}
