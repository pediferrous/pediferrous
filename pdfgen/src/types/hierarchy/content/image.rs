//! Image PDF object types and implementations.

use std::io::{BufReader, Cursor, Error, Read, Write};

use image::ImageReader;
use pdfgen_macros::const_names;

use crate::{
    ObjId,
    types::{
        constants,
        hierarchy::primitives::{name::Name, object::Object, rectangle::Position, unit::Unit},
    },
};

use super::stream::Stream;

/// The colour space in which image samples shall be specified; it can be any type of colour space
/// except Pattern.
#[derive(Debug, Clone, PartialEq, PartialOrd)]
#[allow(dead_code)]
enum ColorSpace {
    /// Device default RGB representation.
    DeviceRgb,

    /// Device default single gray channel representation.
    DeviceGray,
}

impl ColorSpace {
    /// Encode this `ColorSpace` into the given implementor of [`Write`].
    fn write(&self, writer: &mut dyn Write) -> Result<usize, Error> {
        match self {
            ColorSpace::DeviceRgb => Name::new(b"DeviceRGB").write(writer),
            ColorSpace::DeviceGray => Name::new(b"DeviceGray").write(writer),
        }
    }
}

/// Represents the information that should be encoded in the dictionary of an [`Image`] stream.
#[derive(Debug, Clone, PartialEq, PartialOrd)]
struct ImageDict {
    /// The width of the image, in samples.
    width: u32,

    /// The height of the image, in samples.
    height: u32,

    /// The colour space in which image samples shall be specified.
    color_space: ColorSpace,

    /// The number of bits used to represent each colour component. Only a single value shall be
    /// specified; the number of bits shall be the same for all colour components. The value shall
    /// be 1, 2, 4, 8, or (from PDF 1.5) 16. If ImageMask is true, this entry is optional, but if
    /// specified, its value shall be 1.
    bits_per_comp: u8,
}

/// Represents transformations that should be applied to the encoded [`Image`] such as position and
/// scaling.
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
pub struct ImageTransform {
    /// Represents the position of an [`Image`] on the [`Page`].
    ///
    /// [`Page`]: crate::types::hierarchy::page::Page
    pub position: Position,
    /// Represents the scaling of an [`Image`] on the [`Page`].
    ///
    /// [`Page`]: crate::types::hierarchy::page::Page
    pub scale: Position,
}

/// A sampled image (or just image for short) is a rectangular array of sample values, each
/// representing a colour.
#[derive(Debug, PartialEq, PartialOrd)]
pub struct Image {
    /// Raw bytes of the image containing the samples. For an RGB image, each sample is represented
    /// by three values, one for each color component - red, green and blue.
    // NOTE: The source format for an image shall be described by four parameters:
    //       • The width of the image in samples
    //       • The height of the image in samples
    //       • The number of colour components per sample
    //       • The number of bits per colour component
    samples: Stream,

    /// Contains the dictionary for the image encoding the specific information of this image such
    /// as width and height in number of samples.
    dict: ImageDict,

    /// Transformations that should be applied to this image when encoding it into a
    /// [`ContentStream`].
    ///
    /// [`ContentStream`]: super::ContentStream
    transform: ImageTransform,
}

impl Image {
    const_names! {
        SUBTYPE,
        IMAGE,
        WIDTH,
        HEIGHT,
        COLOR_SPACE,
        BITS_PER_COMPONENT,
    }

    /// Creates a new [`Image`] by reading the bytes from the `reader` with default width and
    /// height of 100 mm and position 0, 0 (lower left corner of a page).
    pub fn from_reader(reader: impl Read) -> ImageBuilder<false> {
        let mut bytes = Vec::new();
        BufReader::new(reader).read_to_end(&mut bytes).unwrap();
        Self::from_bytes(bytes)
    }

    pub fn from_file(file: &std::fs::File) -> ImageBuilder<false> {
        let mut bytes = Vec::new();
        BufReader::new(file).read_to_end(&mut bytes).unwrap();
        Self::from_bytes(bytes)
    }

    /// Creates a new [`Image`] from the given bytes with default width and height of 100 mm and
    /// position 0, 0 (lower left corner of a page).
    pub fn from_bytes(bytes: impl Into<Vec<u8>>) -> ImageBuilder<false> {
        let bufreader = Cursor::new(bytes.into());

        let decoded_image = ImageReader::new(bufreader)
            .with_guessed_format()
            .unwrap()
            .decode()
            .unwrap();

        let img = decoded_image.to_rgb8();
        let (width, height) = img.dimensions();
        let pixels = img.into_raw();

        let img = Self {
            samples: Stream::with_bytes(pixels),
            dict: ImageDict {
                width,
                height,
                color_space: ColorSpace::DeviceRgb,
                bits_per_comp: 8,
            },
            transform: ImageTransform {
                position: Position::from_mm(0.0, 0.0),
                scale: Position::from_units(width as f32, height as f32),
            },
        };

        ImageBuilder { inner: img }
    }

    /// Sets the width and height of this [`Image`].
    pub fn set_dimensions(&mut self, width: Unit, height: Unit) {
        self.set_width(width);
        self.set_height(height);
    }

    /// Sets the width of this [`Image`]
    pub fn set_width(&mut self, width: Unit) {
        self.transform.scale.x = width;
    }

    /// Sets the height of this [`Image`]
    pub fn set_height(&mut self, height: Unit) {
        self.transform.scale.y = height;
    }

    /// Sets the position of this [`Image`].
    pub fn set_pos(&mut self, position: Position) {
        self.transform.position = position;
    }

    /// Returns the width, height and position tuple of this [`Image`].
    // TODO: should this be exposed in public API?
    pub fn transform(&self) -> ImageTransform {
        self.transform
    }

    pub fn write(&self, writer: &mut dyn Write, id: &ObjId) -> Result<usize, Error> {
        Ok(pdfgen_macros::write_chain! {
            id.write_def(writer),
            writer.write(constants::NL_MARKER),

            self.write_content(writer),
            self.write_end(writer),
        })
    }
}

impl Object for Image {
    fn write_def(&self, _writer: &mut dyn Write) -> Result<usize, Error> {
        panic!("Image does not fully implement the Object trait.")
    }

    fn write_content(&self, writer: &mut dyn Write) -> Result<usize, Error> {
        // NOTE: The image dictionary shall specify the width, height, and number of bits per
        //       component explicitly. The number of colour components shall be inferred from the
        //       colour space specified in the dictionary.

        Ok(pdfgen_macros::write_chain! {
            self.samples.write_with_dict(writer, |writer| {
                Ok(pdfgen_macros::write_chain! {
                    Name::TYPE.write(writer),
                    Name::X_OBJECT.write(writer),
                    writer.write(constants::NL_MARKER),

                    Self::SUBTYPE.write(writer),
                    Self::IMAGE.write(writer),
                    writer.write(constants::NL_MARKER),

                    Self::WIDTH.write(writer),
                    writer.write(format!("{}", self.dict.width).as_bytes()),
                    writer.write(constants::NL_MARKER),

                    Self::HEIGHT.write(writer),
                    writer.write(format!("{}", self.dict.height).as_bytes()),
                    writer.write(constants::NL_MARKER),

                    Self::COLOR_SPACE.write(writer),
                    self.dict.color_space.write(writer),
                    writer.write(constants::NL_MARKER),

                    Self::BITS_PER_COMPONENT.write(writer),
                    writer.write(format!("{}", self.dict.bits_per_comp).as_bytes()),
                    writer.write(constants::NL_MARKER),
                })
            }),
            writer.write(constants::NL_MARKER),
        })
    }
}

pub struct ImageBuilder<const IS_INIT: bool> {
    inner: Image,
}

impl<const IS_INIT: bool> ImageBuilder<IS_INIT> {
    /// Sets the position of an [`Image`] on a page.
    pub fn at(mut self, pos: Position) -> ImageBuilder<true> {
        self.inner.transform.position = pos;
        ImageBuilder { inner: self.inner }
    }

    /// Sets the scaling of the image to the given width and height.
    pub fn scaled(mut self, scale: Position) -> Self {
        self.inner.transform.scale = scale;
        self
    }

    /// This is not yet implemented and is a no-op for now.
    pub fn rotated(self, _degree: usize) -> Self {
        // TODO: implement rotation
        self
    }
}

impl ImageBuilder<true> {
    pub fn build(self) -> Image {
        self.inner
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use crate::{IdManager, types::hierarchy::primitives::rectangle::Position};

    use super::Image;

    #[test]
    fn sample_image() {
        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("sample_image.jpg");
        let img_file = std::fs::File::open(dbg!(path)).unwrap();

        let mut id_mngr = IdManager::new();

        let img = Image::from_reader(img_file)
            .scaled(Position::from_mm(100., 100.))
            .at(Position::from_mm(10.0, 42.0))
            .build();

        let mut writer = Vec::default();
        // NOTE: same function defined on Image directly, so call using qualified syntax
        img.write(&mut writer, &id_mngr.create_id()).unwrap();
        let output = String::from_utf8_lossy(&writer);

        insta::assert_snapshot!(output);
    }
}
