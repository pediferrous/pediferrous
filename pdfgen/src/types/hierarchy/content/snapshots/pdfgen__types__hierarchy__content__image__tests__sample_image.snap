---
source: pdfgen/src/types/hierarchy/content/image.rs
expression: output
snapshot_kind: text
---
2 0 obj
<< /Type /XObject 
/Subtype /Image 
/Width 64
/Height 64
/ColorSpace /DeviceRGB 
/BitsPerComponent 8
/Length 12288 >>
stream
������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������ynoaXT[UMZYOXYTWZYSVV[[[ZXR]YPa]UWWP[`\PWV������������������������������������������������������������������������������������������������������������������������������������������������T;G;%C%4R6-G4('1./\^c("%|faM5,sibDE@)1,!#:9?���������������������������������������������������������������������������������������������������������������������������������������8*0F 0f2C�HUi39_=6iTIL@:Ļ�dX]hV\z_^¬����>?<x}{���XNR2&+���������������������������������������������������������������������������������������������������������������������������������l]hM1@|O\�dmn.1���7@71���E:7;�hj]LMslnNHK���0(,'���ϯ�=":&������������������������������������������������������������������������������������������������������������������������������* Y:Dp@Fz@Cm-.���C//�ÿ``YXNGpQO`EEvtxdnt{��HGPpel'Ơ�ŕ�yELa7<C/3���������������������������������������������������������������������������������������������������������������������������1)!D,#�PL�RR���G"���USM|xp���LA>>==Pdk���Rlt,9BEAF,ə��|��RYҘ�20)+������������������������������������������������������������������������������������������������������������������������&u]SU+%�]\J$(׻�H;9:50������]dc^im���!-l}�r��$+* ͟�ʄ��X_���̏�E%=,5���������������������������������������������������������������������������������������������������������������������A50gPOY@>YDD̼�PCE������>C?RccZuwy��.
xL\/%sw}ds� /���墢�>G�MZ�O[�'3N!;+������������������������������������������������������������������������������������������������������������������&P>Bxccĭ�5&XCJ������&13hz}{��/$/�VhY+pIT<+6ij}.���嶵�cj�[g�ku�eq�}�dGT.#,���������������������������������������������������������������������������������������������������������������|wv���T&+_%4=udm\_gar|y�7~KUX+:6
O.vHYK+;&dY`������������������¾�ACJY_e������������������������������������������������������������������������������������������������������������>HN�~�[13Ԓ��^mZ"5*!-u~�ru�1-�PZ_&0eFN���zP_P 0wJW@'&%'%"&''.������������������������������������������������������������������������������������������������������������$4&.j18�jm�U\I(}��w~�E$0}NYR-7C+1���������S=BU08mDMJ'4SARen�$%:V,8���uDN5$#*7'1������������������������������������������������������������������������������������������������������������-5H���J!*l6<S'/�oy�y�-!.xN\R(6:-7TV[���674������5%*G/8^;JG*?hd~���H"2n=II"/[HT{��'2������������������������������������������������������������������������������������������������������������*W\j��� &/lv����G+vIX?+;+���0:foo��߸��n��#.:=+:`BRN7Dk\j��'4:KX���)/8 &������������������������������������������������������������������������������������������������������������0!,8,9gn{���z��/#0p?Kc6C���u�����%//,-cTPF97g��x�����bP^dEMQ;Bjq|u�����-4A:/6.#������������������������������������������������������������������������������������������������������������C�R\A 3`CV<!g=II*45*5���<J���6:@߿�䳭���]NJfx{Iiq[t�_fs<'/`DK7)1}jxM%8xHX_6=4������������������������������������������������������������������������������������������������������������Dh49O#+;Y )c4;F9Abox���5N[>AKҺ�ﵸ�XZ֖�廽wtv6HO������JAL?)0H124S%i3;M$)/������������������������������������������������������������������������������������������������������������"N96ʨ�ĝ�Ȣ�iSPHNT���6MUX[cӬ����v(9�DX�5I�~���~s|<EScq�qw�!"���Ū�����xl" ������������������������������������������������������������������������������������������������������������992��Ʃ���s`��|�ǻS_\9KN#,,Ȳ�׎��'7j 9j">r:�+F�Q`��`R`/4@HOX���Ƶ��v^��s�ͻ��� "������������������������������������������������������������������������������������������������������������M:,׿�S@4\K<B/������75/���ک��gn�:GT,=ĭ�}Qkv0I�E[�|���NC@��u��ĀveTC4K3)������&������������������������������������������������������������������������������������������������������������>+��o6#ĳ�F0�~g���d_Z���ѝ��W[�<EX,;ͳńWnx1I�BS�nwٳ�PF@���º�F8&���tXMqUL�lg%������������������������������������������������������������������������������������������������������������GG@���w`Q]=)pN9�˻���1>E" "å�ʂ��+5u#0v#5&:�+9�OO�ÿthm(06"/.��٨�W6`@(Һ����������������������������������������������������������������������������������������������������������������SJF���ֳ�����mfILT}��(@<E�������(4�8E�.;�z���ѓ@Q\��"!���侪�˵���($��������������������������������������������������������������������������������������������������������󸺰#F67VA?P1/`23K%)>3>q{����l��elw���⥫�EF������ubnmk{������]Y^3^;:_;6U;5RB;B:2������������������������������������������������������������������������������������������������������������8++����mqM),R&wDNG%43+;���(:���&,5���͢��ľ`[Zpu�:EZSfwbgs8 &f>AX.2W.3U16¦����$������������������������������������������������������������������������������������������������������������B5={ckL&*U#'i,2S�ER_9K���������*6/(2m`hSOV(j��������yhssDKi/2^++U#k*3i3;���$������������������������������������������������������������������������������������������������������������)#I.7R).Ȗ��]^n.1Y�DTF.?)���,6s}���Γ��'t��'2<@3?kFRh+5i(-g64߭�d#*h.5=!-$������������������������������������������������������������������������������������������������������������4'U8C�kuP25̬�Jr%-[�WbA(IAGX_b���T\Y�����}K@?D-.~TYg,6m$0^$d?=�hb���J%%Z69.������������������������������������������������������������������������������������������������������������."='.Ĳ����>.'Z:3h',v+3R�X\K-0G:<���������]QKK.+~UTNj-5f!.b)2<#������fKH<.������������������������������������������������������������������������������������������������������������,%#���LBEcY\���@,,pAH\(/^9;>�ksN"*XGH���r^`B$'�[\J! wOPG!vCKG!�����vJ8-{g_SD@$������������������������������������������������������������������������������������������������������������20,HA?whf`OPdYZ˻�E&,G'+θ����A
�ZdL#(O-2K"+mBJP&(�zz׼�}dc0����}t\MAoZPYE?RGD������������������������������������������������������������������������������������������������������������)""r][lEC�lk�xzV?Bȶ�tflVIQ������C�QW^$x9DV)kCFε�7,){rp����pn}YS��z�gaV96XDF% ������������������������������������������������������������������������������������������������������������!nUS�llk23�]^���]RS���SOV���F (b,2N�FI[n68Z74q^Z�{y���}rsnWV�mii84�nl�rs?+/QIM������������������������������������������������������������������������������������������������������������MJKS><�ST�z}f(,�qvB6:WZ\���:(,r?Gh'.d,&Tp6+[) �e[2���xlqD28tZ\lGG���jBE�qu."������������������������������������������������������������������������������������������������������������������(hGIX(.�S[���_SWgjlMPP̿�U)1^(.�e_����sf_>0H6*���oehUHPWEL�io�qsP/0�ilmW\9-2���������������������������������������������������������������������������������������������������������������������$2 &|dm)zlrFBG+-2MKO���_HKyhg���{v]SMUVShml%$,JCMhZbC.5oUWO::C891.0���������������������������������������������������������������������������������������������������������������������������zx~81;"$" #$!%)!"" $##!  # &!%"%2"#OBB��������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������鎝������롯���������������������������������钞������鮻Ƒ����������������������Ɠ������������������钜����������������������������������������������������������������������������_k|���Tcy�����������Uh���Ugz������APd�����_n������뢶�q��������bt����Rdw���9Lc���AUi���Xk����p��}����Ț�����]iz��Ā��������������������������������������������������������������������������Q_r|��]o����Law`w����Qf~���s��������DXl������EYmr����ә��I^vex����`s����>Si���,C^���H_y���k�����bv�����і� �ɉ��OZn���������������������������������������������������������������������������Xdu��������̄��������Qf|���n��������@Re������Ui{���������{��������`s�}��`u����6Jer����Ⱥ��m�����ez�����ɖ�����������Waq������������������������������������������������������������������������ry����������HVgZj|���VeyGUfZiw���ix�;GXq}����v�����������CQd]k|���ds����FVf���>Ma���Pcr���M\pM\p������EUeDTf���XcpS]g������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������������
endstream
endobj
